### PR TITLE
Add Graphite support as an alternative to InfluxDB

### PR DESCRIPTION
**Pull request by bartek-wesolowski**

### CODE REVIEW

**Comment by bartek-wesolowski** [comment]
Linked issue: https://github.com/Malinskiy/marathon/issues/400

File: cli/src/main/kotlin/com/malinskiy/marathon/cli/config/deserialize/GraphiteConfigurationDeserializer.kt
Comment (Malinskiy) [inline cli/src/main/kotlin/com/malinskiy/marathon/cli/config/deserialize/GraphiteConfigurationDeserializer.kt]: Deserialiser is not responsible for handling default values. Please make the port optional (as in nullable) and handle the defaults outside, e.g. in the data class definition via constructor default argument value.

File: cli/src/main/kotlin/com/malinskiy/marathon/cli/config/deserialize/GraphiteConfigurationDeserializer.kt
Comment (Malinskiy) [inline cli/src/main/kotlin/com/malinskiy/marathon/cli/config/deserialize/GraphiteConfigurationDeserializer.kt:23]: The only reason AFAIU for this deserialiser to exist is custom error handling. Correct me if I'm wrong. Thanks.

File: core/src/main/kotlin/com/malinskiy/marathon/analytics/TrackerFactory.kt
Comment (Malinskiy) [inline core/src/main/kotlin/com/malinskiy/marathon/analytics/TrackerFactory.kt:43]: Would be easier to read if it was done using when, e.g.:

```kotlin
        val analyticsConfiguration = configuration.analyticsConfiguration
        val analyticsTracker = when (analyticsConfiguration) {
            is InfluxDbConfiguration -> createInfluxDbTracker(analyticsConfiguration)
            is GraphiteConfiguration -> createGraphiteTracker(analyticsConfiguration)
            else -> null
        }
        if (analyticsTracker != null) {
            defaultTrackers.add(analyticsTracker)
        }
```

File: core/src/main/kotlin/com/malinskiy/marathon/analytics/external/MetricsProviderFactory.kt
Comment (Malinskiy) [inline core/src/main/kotlin/com/malinskiy/marathon/analytics/external/MetricsProviderFactory.kt:17]: Even if the configuration is correct, there will be a case of network failure. The NoOpMetricsProvider fallback is there for a reason: when the graphite server will fail - we will fallback. The Influx implementation just does `try/catch` to verify everything is working. You should implement the same idea for graphite here.

File: core/src/main/kotlin/com/malinskiy/marathon/analytics/external/graphite/GraphiteUtils.kt
Comment (Malinskiy) [inline core/src/main/kotlin/com/malinskiy/marathon/analytics/external/graphite/GraphiteUtils.kt]: We have a convention to put every extension of default classes into `extension` package. Could you please move this to `com.malinskiy.marathon.extension`?

File: core/src/main/kotlin/com/malinskiy/marathon/analytics/external/graphite/BasicGraphiteClient.kt
Comment (Malinskiy) [inline core/src/main/kotlin/com/malinskiy/marathon/analytics/external/graphite/BasicGraphiteClient.kt:20]: All 3 instantiations of GraphiteMetric provide the timestamp value. Do we really need a default here?

File: core/src/main/kotlin/com/malinskiy/marathon/analytics/external/graphite/GraphiteTracker.kt
Comment (Malinskiy) [inline core/src/main/kotlin/com/malinskiy/marathon/analytics/external/graphite/GraphiteTracker.kt:14]: Graphite supports [tags](https://graphite.readthedocs.io/en/latest/tags.html)

We should use them to allow constructing dashboards and querying data in a simple manner. See the InfluxDbTracker to get the idea of the tags that you need to pass.

File: core/src/main/kotlin/com/malinskiy/marathon/analytics/metrics/remote/graphite/GraphiteDataSource.kt
Comment (Malinskiy) [inline core/src/main/kotlin/com/malinskiy/marathon/analytics/metrics/remote/graphite/GraphiteDataSource.kt:51]: The current implementation of influx is addressing the fact that we query a lot of data from influx: it reuses the connection. This implementation of graphite will slow down tests. Is there any way we can reuse the connect and skip the TCP handshake for potentially thousands of requests?

File: core/src/main/kotlin/com/malinskiy/marathon/analytics/metrics/remote/graphite/GraphiteDataSource.kt
Comment (Malinskiy) [inline core/src/main/kotlin/com/malinskiy/marathon/analytics/metrics/remote/graphite/GraphiteDataSource.kt]: We use okhttp for http connections, I suggest you rewrite this so as to not introduce another HTTP stack to the project.

File: cli/src/main/kotlin/com/malinskiy/marathon/cli/config/deserialize/AnalyticsConfigurationDeserializer.kt
Comment (Malinskiy) [inline cli/src/main/kotlin/com/malinskiy/marathon/cli/config/deserialize/AnalyticsConfigurationDeserializer.kt]: It's possible to specify both configurations at the same time. We should produce an error that it's incorrect to provide both influx and graphite at the same time. Or at the very least, produce a warning that we influx configuration will override the graphite one.

**Review by Malinskiy — CHANGES_REQUESTED** [review]
Good work! I've noticed some things that need some work in the PR and commented on those.
Apart from the requested changes this PR also needs:
- [ ] Integration test with graphite similar to InfluxDbProviderIntegrationTest to verify the initialisation of remote data source
- [ ] Integration test for the data source (similar to InfluxDBDataSourceTest)
- [ ] Add graphite test case in MetricsProviderFactoryTest to test when config is correct and also when it's not correct. Including fallback to NoOp when graphite server is down/unavailable.
- [ ] Add configuration documentation to the `gh-pages` branch in a separate PR so that our users know about this new option.

Also please work on the code coverage until your diff is at least not decreasing the coverage. We can discuss if you'll see some roadblocks on the way, but generally, we don't merge new code that is not well tested.

File: core/src/main/kotlin/com/malinskiy/marathon/analytics/external/graphite/BasicGraphiteClient.kt
Comment (bartek-wesolowski) [reply core/src/main/kotlin/com/malinskiy/marathon/analytics/external/graphite/BasicGraphiteClient.kt:20]: I copied the code of `BasicGraphiteClient` from [Graphite Kotlin Client](https://github.com/mihkels/graphite-client-kotlin). Unfortunately, the library is not available on Maven Central yet. I'd rather not modify the code, because then it can be easily replaced with the library implementation in the future.

I tried using [graphite-client](https://github.com/savantly-net/graphite-client) before, but for some reason it wasn't sending or fetching any data.

File: core/src/main/kotlin/com/malinskiy/marathon/analytics/external/MetricsProviderFactory.kt
Comment (bartek-wesolowski) [reply core/src/main/kotlin/com/malinskiy/marathon/analytics/external/MetricsProviderFactory.kt:17]: `GraphiteDataSource` never throws any exceptions in its constructor. It doesn't create any database connections. Data is fetched on demand via HTTP requests. `createGraphiteMetricsProvider` will always succeed and it doesn't need a fallback.

File: core/src/main/kotlin/com/malinskiy/marathon/analytics/metrics/remote/graphite/GraphiteDataSource.kt
Comment (bartek-wesolowski) [reply core/src/main/kotlin/com/malinskiy/marathon/analytics/metrics/remote/graphite/GraphiteDataSource.kt:51]: Fetching data from Graphite is done via HTTP requests. No database connection is kept open and consequently it doesn't have to be closed. This approach shouldn't slow down tests, because one single query, either for all success rates or for all execution times, is executed before the tests are run and the query takes a fraction of a second.

File: cli/src/main/kotlin/com/malinskiy/marathon/cli/config/deserialize/GraphiteConfigurationDeserializer.kt
Comment (bartek-wesolowski) [reply cli/src/main/kotlin/com/malinskiy/marathon/cli/config/deserialize/GraphiteConfigurationDeserializer.kt:23]: To be honest, I used `InfluxDbConfigurationDeserializer` as an example and created a similar implementation for Graphite. But yes, it displays detailed error logs when the configuration is incorrect. Do you think this should be handled differently?

File: cli/src/main/kotlin/com/malinskiy/marathon/cli/config/deserialize/AnalyticsConfigurationDeserializer.kt
Comment (bartek-wesolowski) [reply cli/src/main/kotlin/com/malinskiy/marathon/cli/config/deserialize/AnalyticsConfigurationDeserializer.kt]: Done. Added throwing a `ConfigurationException` when both InfluxDB and Graphite configurations are specified.

File: core/src/main/kotlin/com/malinskiy/marathon/analytics/metrics/remote/graphite/GraphiteDataSource.kt
Comment (bartek-wesolowski) [reply core/src/main/kotlin/com/malinskiy/marathon/analytics/metrics/remote/graphite/GraphiteDataSource.kt]: done

File: core/src/main/kotlin/com/malinskiy/marathon/analytics/external/graphite/GraphiteTracker.kt
Comment (bartek-wesolowski) [reply core/src/main/kotlin/com/malinskiy/marathon/analytics/external/graphite/GraphiteTracker.kt:14]: I tried using tags but they didn't work for me. It turned out that my company disabled them for some technical reason. I'm not able to implement tags right now, because I can't test them. Also, I wouldn't be able to use the Graphite integration if it used tags. I'd rather not change this now and perhaps add tags later as an option.

File: core/src/main/kotlin/com/malinskiy/marathon/analytics/external/graphite/GraphiteUtils.kt
Comment (bartek-wesolowski) [reply core/src/main/kotlin/com/malinskiy/marathon/analytics/external/graphite/GraphiteUtils.kt]: done

File: core/src/main/kotlin/com/malinskiy/marathon/analytics/TrackerFactory.kt
Comment (bartek-wesolowski) [reply core/src/main/kotlin/com/malinskiy/marathon/analytics/TrackerFactory.kt:43]: done

File: cli/src/main/kotlin/com/malinskiy/marathon/cli/config/deserialize/GraphiteConfigurationDeserializer.kt
Comment (bartek-wesolowski) [reply cli/src/main/kotlin/com/malinskiy/marathon/cli/config/deserialize/GraphiteConfigurationDeserializer.kt]: done

**Comment by bartek-wesolowski** [comment]
As to the unit tests:
- IMHO a test similar to `InfluxDbProviderIntegrationTest` is not necessary, because Graphite doesn't have a class like `GraphiteProvider` and `BasicGraphiteClient'`s constructor invocation will never fail.
- I added unit tests for `GraphiteDataSource`
- IMHO additional tests in `MetricsProviderFactoryTest` aren't necessary, because creating `GraphiteDataSource` can never fail

**Comment by bartek-wesolowski** [comment]
Add a PR with documentation:
https://github.com/Malinskiy/marathon/pull/408

**Comment by bartek-wesolowski** [comment]
Could someone please add a "hacktoberfest-accepted" label to this PR so that I can take part in Hacktoberfest 2020 with these changes?

**Comment by tagantroy** [comment]
> * IMHO a test similar to `InfluxDbProviderIntegrationTest` is not necessary, because Graphite doesn't have a class like `GraphiteProvider` and `BasicGraphiteClient'`s constructor invocation will never fail.

I think the point of the integration test - verify the complete integration with a real database. We caught a couple of issues with InfluxDB using this approach.
I think we can merge this diff after the integration test is added.
Thank you for your contribution! :)

**Comment by bartek-wesolowski** [comment]
> > * IMHO a test similar to `InfluxDbProviderIntegrationTest` is not necessary, because Graphite doesn't have a class like `GraphiteProvider` and `BasicGraphiteClient'`s constructor invocation will never fail.
> 
> I think the point of the integration test - verify the complete integration with a real database. We caught a couple of issues with InfluxDB using this approach.
> I think we can merge this diff after the integration test is added.
> Thank you for your contribution! :)

I added another unit test for `BasicGraphiteClient`.

As to the integration test, I don't think I'm able to write it. Opposed to InfluxDB, you can't start a Graphite database from code. The Graphite library that I use just wraps Socket and HTTP connections and it consists of one short file. The unit tests already check whether correct data is sent/fetched.

**Comment by tagantroy** [comment]
May I ask you share some examples of dashboard?(for example https://github.com/Malinskiy/marathon/blob/develop/assets/grafana-dashboard.json) 
I created template for integration test(patch attached)
[patch.txt](https://github.com/Malinskiy/marathon/files/5418714/patch.txt)

**Comment by bartek-wesolowski** [comment]
> May I ask you share some examples of dashboard?(for example https://github.com/Malinskiy/marathon/blob/develop/assets/grafana-dashboard.json)
> I created template for integration test(patch attached)
> [patch.txt](https://github.com/Malinskiy/marathon/files/5418714/patch.txt)

I added a sample dashboard for Graphite. I'll see what I can do about the integration test. The sample dashboard looks like this with some sample data.

Zrzut ekranu 2020-10-23 o 08 18 42

File: cli/src/main/kotlin/com/malinskiy/marathon/cli/config/deserialize/GraphiteConfigurationDeserializer.kt
Comment (Malinskiy) [reply cli/src/main/kotlin/com/malinskiy/marathon/cli/config/deserialize/GraphiteConfigurationDeserializer.kt:23]: No, that's fine, leave it as is. 
We're thinking of moving to something else in terms of configuration parsing so that we get proper error messages out-of-the-box instead of writing custom deserialisers. The whole point was to define data model and let the parse handle everything, but json/yaml packages lack proper error handling for our particular case.

File: core/src/main/kotlin/com/malinskiy/marathon/analytics/external/MetricsProviderFactory.kt
Comment (Malinskiy) [reply core/src/main/kotlin/com/malinskiy/marathon/analytics/external/MetricsProviderFactory.kt:17]: That might be sub-optimal in terms of performance (we might request thousands of metrics throughout the execution of a large test suite). For starters, this should be fine though. No need to change.

File: core/src/main/kotlin/com/malinskiy/marathon/analytics/external/graphite/BasicGraphiteClient.kt
Comment (Malinskiy) [reply core/src/main/kotlin/com/malinskiy/marathon/analytics/external/graphite/BasicGraphiteClient.kt:20]: gotcha. Can you leave a comment there that this is not original code and leave a link to source? thanks

File: core/src/main/kotlin/com/malinskiy/marathon/analytics/external/graphite/GraphiteTracker.kt
Comment (Malinskiy) [reply core/src/main/kotlin/com/malinskiy/marathon/analytics/external/graphite/GraphiteTracker.kt:14]: They're disabled probably because of high cardinality

File: core/src/main/kotlin/com/malinskiy/marathon/analytics/external/graphite/BasicGraphiteClient.kt
Comment (bartek-wesolowski) [reply core/src/main/kotlin/com/malinskiy/marathon/analytics/external/graphite/BasicGraphiteClient.kt:20]: done

**Comment by bartek-wesolowski** [comment]
I added an integration test for Graphite and refactored integration tests a little bit so that the test cases are shared between InfluxDb and Graphite tests.

**Review by tagantroy — APPROVED** [review]
I'm not a big fan of inheritance in tests, but I think it is ok in this case.
LGTM!

**Comment by bartek-wesolowski** [comment]
@Malinskiy check it out once again please. Your comments have been addressed.

**Comment by bartek-wesolowski** [comment]
Guys, would you mind taking a look at this PR once again? Do you think it can be merged now or does it still need some work?

**Review by tagantroy — APPROVED** [review]

**Pull request merged**
